### PR TITLE
CORE-3775: Update the test lifecycle coordinator code and add a test

### PR DESCRIPTION
**Pull request by fowlerrr**

### CODE REVIEW

**Review by JamesHR3 — COMMENTED** [review]
Seems reasonable, one minor suggestion.

File: libs/lifecycle/lifecycle-impl/build.gradle [hunk 1]
```diff
@@ -18,6 +18,7 @@ dependencies {
     implementation 'org.slf4j:slf4j-api'
 
     testImplementation project(":testing:test-utilities")
+    testImplementation project(":libs:lifecycle:lifecycle-impl")
```
Comment (JamesHR3) [inline libs/lifecycle/lifecycle-impl/build.gradle]: Possibly I'm missing something, but I think you're declaring a dependency on yourself here?

File: libs/lifecycle/lifecycle-test-impl/src/main/kotlin/net/corda/lifecycle/test/impl/TestLifecycleCoordinatorFactoryImpl.kt [hunk 2]
```diff
@@ -1,21 +1,14 @@
...
-        return coordinators.compute(name) { _, _ ->
-            TestLifecycleCoordinatorImpl(name, handler)
-        } as LifecycleCoordinator
-    }
-}
+import net.corda.lifecycle.impl.LifecycleCoordinatorFactoryImpl
+import net.corda.lifecycle.impl.registry.LifecycleRegistryCoordinatorAccess
+import net.corda.lifecycle.impl.registry.LifecycleRegistryImpl
+
+class TestLifecycleCoordinatorFactoryImpl(
```
Comment (JamesHR3) [inline libs/lifecycle/lifecycle-test-impl/src/main/kotlin/net/corda/lifecycle/test/impl/TestLifecycleCoordinatorFactoryImpl.kt:8]: Wonder if we might want to OSGi inject this at some later point? No need to do that in this PR though.

File: libs/lifecycle/lifecycle/src/main/kotlin/net/corda/lifecycle/LifecycleCoordinatorScheduler.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package net.corda.lifecycle
+
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+
+interface LifecycleCoordinatorScheduler {
```
Comment (JamesHR3) [inline libs/lifecycle/lifecycle/src/main/kotlin/net/corda/lifecycle/LifecycleCoordinatorScheduler.kt:11]: This should probably have some kdocs.

Possibly we should actually put it in the lifecycle impl, as we're not expecting the user of the library to ever give us one of these or to ever implement the interface. There's an interesting question around OSGi if we do that but again, not a problem for this PR.

File: libs/lifecycle/lifecycle/src/main/kotlin/net/corda/lifecycle/LifecycleCoordinatorSchedulerFactory.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package net.corda.lifecycle
+
+interface LifecycleCoordinatorSchedulerFactory {
```
Comment (JamesHR3) [inline libs/lifecycle/lifecycle/src/main/kotlin/net/corda/lifecycle/LifecycleCoordinatorSchedulerFactory.kt:9]: Same thing here about kdocs.

File: libs/lifecycle/lifecycle-test-impl/src/main/kotlin/net/corda/lifecycle/test/impl/LifecycleTest.kt [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        verificationWhenDown: () -> Unit = {},
+        verificationWhenUp: () -> Unit = {},
+    ) {
+        bringDependencyDown<T>()
+        verifyIsDown<T>()
+        verificationWhenDown.invoke()
+        bringDependencyUp<T>()
+        verifyIsUp<T>()
+        verificationWhenUp.invoke()
+    }
```
Comment (fowlerrr) [inline libs/lifecycle/lifecycle-test-impl/src/main/kotlin/net/corda/lifecycle/test/impl/LifecycleTest.kt:112]: I was playing about and this method snuck in.  We can remove it.  But it allows someone to turn off/then on again a dependency and verify state, if they wish.  Just hiding some of the lines that might happen a lot.  Not sure if it's worth saving for later, though...

File: components/flow/flow-service/build.gradle [hunk 6]
```diff
@@ -52,6 +52,7 @@ dependencies {
 
     testImplementation project(":libs:flows:session-manager-impl")
     testImplementation project(":libs:lifecycle:lifecycle-test-impl")
+    testImplementation project(":libs:lifecycle:lifecycle-impl")
```
Comment (fowlerrr) [inline components/flow/flow-service/build.gradle:55]: Can be saved for a follow up, but it'd be nice if we could avoid _needing_ `lifecycle-impl` in addition to `lifecycle-test-impl`.  Could we do something in the `lifecycle-test-impl` gradle to package it?  IIRC, isn't that what `api` does?

**Review by JamesHR3 — APPROVED** [review]

**Review by owenstanford — APPROVED** [review]

**Pull request merged**